plugins/h3/h3.cpp

```diff
@@ -0,0 +1,247 @@
+}
+
+namespace h3
+{
+
+//--------------------------------------------------------------------------------
+//                           ECL SERVICE ENTRYPOINTS
+//--------------------------------------------------------------------------------
+
+//  Indexing Functions ---
+ECL_H3_API unsigned __int64 ECL_H3_CALL index(ICodeContext *_ctx, double lat, double lon, uint32_t resolution)
+{
+    GeoCoord location;
+    location.lat = ::degsToRads(lat);
+    location.lon = ::degsToRads(lon);
+    return ::geoToH3(&location, resolution);
+}
+
+ECL_H3_API void ECL_H3_CALL center(ICodeContext *_ctx, size32_t &__lenResult, void *&__result, unsigned __int64 index)
+{
+    GeoCoord location;
+    ::h3ToGeo(index, &location);
+    __lenResult = 2 * sizeof(double);
+    __result = rtlMalloc(__lenResult);
+    double *cur = static_cast<double *>(__result);
+    *cur = ::radsToDegs(location.lat);
+    *(++cur) = ::radsToDegs(location.lon);
+}
+
+ECL_H3_API void ECL_H3_CALL boundary(ICodeContext *_ctx, size32_t &__lenResult, void *&__result, unsigned __int64 index)
+{
+    GeoBoundary boundary;
+    ::h3ToGeoBoundary(index, &boundary);
+    __lenResult = boundary.numVerts * 2 * sizeof(double);
+    __result = rtlMalloc(__lenResult);
+
+    double *cur = static_cast<double *>(__result);
+    for (int v = 0; v < boundary.numVerts; v++)
+    {
+        *cur = ::radsToDegs(boundary.verts[v].lat);
+        *(++cur) = ::radsToDegs(boundary.verts[v].lon);
+        ++cur;
+    }
+}
+
+//  Introspection ---
+ECL_H3_API uint32_t ECL_H3_CALL resolution(ICodeContext *_ctx, unsigned __int64 index)
+{
+    return ::h3GetResolution(index);
+}
+
+ECL_H3_API uint32_t ECL_H3_CALL baseCell(ICodeContext *_ctx, unsigned __int64 index)
+{
+    return ::h3GetBaseCell(index);
+}
+
+ECL_H3_API void ECL_H3_CALL toString(ICodeContext *_ctx, size32_t &lenVarStr, char *&varStr, unsigned __int64 index)
+{
+    varStr = static_cast<char *>(rtlMalloc(17));
+    ::h3ToString(index, varStr, 17);
+    lenVarStr = strlen(varStr);
+}
+
+ECL_H3_API unsigned __int64 ECL_H3_CALL fromString(ICodeContext *_ctx, const char *strIdx)
+{
+    return ::stringToH3(strIdx);
+}
+
+ECL_H3_API bool ECL_H3_CALL isValid(ICodeContext *_ctx, unsigned __int64 index)
+{
+    return ::h3IsValid(index) != 0;
+}
+
+//  Traversal  ---
+ECL_H3_API void ECL_H3_CALL kRing(ICodeContext *_ctx, bool &__isAllResult, size32_t &__lenResult, void *&__result, unsigned __int64 index, int32_t k)
+{
+    int maxBuff = ::maxKringSize(k);
+    H3Index *buff = static_cast<H3Index *>(calloc(maxBuff, sizeof(H3Index)));
+    ::kRing(index, k, buff);
+    toSetOf(__isAllResult, __lenResult, __result, buff, maxBuff);
+    free(buff);
+}
+
+ECL_H3_API void ECL_H3_CALL hexRing(ICodeContext *_ctx, bool &__isAllResult, size32_t &__lenResult, void *&__result, unsigned __int64 index, uint32_t k)
+{
+    int maxBuff = ::maxKringSize(k);
+    H3Index *buff = static_cast<H3Index *>(calloc(maxBuff, sizeof(H3Index)));
```

Comment:
I don't know if exceptions are a possibility but if they are buff is likely to leak. Use of a rtlDataAttr may be preferable.

Also seems a shame to have to copy the data twice - is this to eliminate zeros?

Reply:
@richardkchapman - Given H3 is a "c" library they should not be throwing any explicit exceptions - is it still wise to use rtlDataArr?
Yes unfortunately the data can come back "gappy" and either we remove the null records here, or force all the consumers to be aware of that.

Reply:
You actually copy the data 3 times I think - once into buff, then from there into a vector, then from there into the set that you return.

How common are the gaps, and how often are there less than maxBuff elements added to the ring? It may be preferable to get the data directly into the target set and then close up gaps.

Or this whole discussion may come under the "premature optimization" banner.

Reply:
Agree - testing simpler version now.

Reply:
@richardkchapman ready for re-review